src/storage/src/hummock/shared_buffer/shared_buffer_uploader.rs

```diff
@@ -140,68 +203,21 @@ impl SharedBufferUploader {
         // Ensure the added data is available locally
         self.local_version_manager.try_set_version(version);
 
-        Ok(sync_size)
-    }
-
-    async fn handle(&mut self, item: SharedBufferUploaderItem) -> StorageResult<()> {
-        match item {
-            SharedBufferUploaderItem::Batch(m) => {
-                if let Some(detector) = &self.write_conflict_detector {
-                    detector.check_conflict_and_track_write_batch(&m.inner, m.epoch);
-                }
-
-                self.batches_to_upload
-                    .entry(m.epoch())
-                    .or_insert(Vec::new())
-                    .push(m);
-                Ok(())
-            }
-            SharedBufferUploaderItem::Sync(sync_item) => {
-                let res = match sync_item.epoch {
-                    Some(e) => {
-                        // Sync a specific epoch
-                        self.sync(e).await
-                    }
-                    None => {
-                        // Sync all epochs
-                        let epochs = self.batches_to_upload.keys().copied().collect_vec();
-                        let mut res = Ok(0);
-                        let mut size_total: u64 = 0;
-
-                        for e in epochs {
-                            res = self.sync(e).await;
-                            if res.is_err() {
-                                break;
-                            }
-                            size_total += res.unwrap();
-                            res = Ok(size_total);
-                        }
-                        res
-                    }
-                };
-
-                if let Some(tx) = sync_item.notifier {
-                    tx.send(res).map_err(|_| {
-                        HummockError::shared_buffer_error(
-                            "Failed to notify shared buffer sync because of send drop",
-                        )
-                    })?;
-                }
-                Ok(())
-            }
-            SharedBufferUploaderItem::Reset(epoch) => {
-                self.batches_to_upload.remove(&epoch);
-                Ok(())
+        // Release shared buffer.
+        let mut removed_size = 0;
+        let mut guard = self.core.write();
+        let epoch_buffer = guard.buffer.get_mut(&epoch).unwrap();
+        for batch in batches {
+            if let Some(removed_batch) = epoch_buffer.remove(batch.end_user_key()) {
+                removed_size += removed_batch.size as usize;
```

Comment:
Seems we cannot remove flushed batches immediately.

https://github.com/singularity-data/risingwave/blob/main/src/storage/src/hummock/state_store.rs#L173-L192

Considering this case:

```plain
a.1 get => max committed epoch < get epoch
b.1 flush => try set version
b.2 flush => remove batches in shared buffer
a.2 get => get nothing, BUG!!
```

Reply:
In the original implementation, it also didn't remove flushed batches. It only subtracted the stat size.

Reply:
Let me think twice if this can happen, 🤔.

Reply:
Any comments? @hzxa21  @zwang28 @Little-Wallace

Reply:
My hunch is that shared buffer should be part of the local version

Reply:
Plus, in https://github.com/singularity-data/risingwave/blob/main/src/storage/src/hummock/state_store.rs#L173-L192 , L180 always read ssts if epoch to get is larger than max committed epoch.


```rust 
    fn get<'a>(&'a self, key: &'a [u8], epoch: u64) -> Self::GetFuture<'_> {
        async move {
            let version = self.local_version_manager.get_version()?;
            // check epoch validity
            validate_epoch(version.safe_epoch(), epoch)?;

            // Query shared buffer. Return the value without iterating SSTs if found
            if version.max_committed_epoch() < epoch {
                if let Some(v) = self
                    .shared_buffer_manager
                    .get(key, (version.max_committed_epoch() + 1)..=epoch)
                {
                    self.stats.get_shared_buffer_hit_counts.inc();
                    return Ok(v.into_user_value().map(|v| v.into()));
                }
            }
            let internal_key = key_with_epoch(key.to_vec(), epoch);

            let mut table_counts = 0;
            for level in &version.levels() {
```

Reply:
If we always read shared buffer then ssts, the visibility will be hard to control.

Reply:
> Seems we cannot remove flushed batches immediately.

We cannot. We only remove flushed batch whose epoch is less than all pinned local versions' max committed epoch, when unpinning a local version (`delete_before`).

Reply:
> We cannot. We only remove flushed batch whose epoch is less than all pinned local versions' max committed epoch, when unpinning a local version (`delete_before`).

But we actually need it to prevent shared buffer OOM and not cause write stall too frequently, QwQ.